failpoints/io_error_test.rs

```diff
@@ -0,0 +1,336 @@
+            write_tmp_engine(ReadableSize::kb(1024), ReadableSize::kb(4), 1024, 1, 4)
+        })
+        .is_err());
+
+        // panic when sync
+        fail::cfg("active_file::truncate::force", "off").unwrap();
+        fail::cfg("log_fd::truncate::err", "off").unwrap();
+        fail::cfg("log_fd::sync::err", "return").unwrap();
+        assert!(panic::catch_unwind(|| {
+            write_tmp_engine(ReadableSize::kb(1024), ReadableSize::kb(4), 1024, 1, 4)
+        })
+        .is_err());
+
+        // panic when create file
+        fail::cfg("log_fd::sync::err", "off").unwrap();
+        fail::cfg("log_fd::create::err", "return").unwrap();
+        assert!(panic::catch_unwind(|| {
+            write_tmp_engine(ReadableSize::kb(1024), ReadableSize::kb(4), 1024, 1, 4)
+        })
+        .is_err());
+
+        // panic when write header
+        fail::cfg("log_fd::create::err", "off").unwrap();
+        fail::cfg("log_fd::write::err", "return").unwrap();
+        assert!(panic::catch_unwind(|| {
+            write_tmp_engine(ReadableSize::kb(1024), ReadableSize::kb(4), 1024, 1, 4)
+        })
+        .is_err());
+        fail::cfg("log_fd::write::err", "off").unwrap();
+    }
+
+    type Action = dyn FnOnce() + Send + Sync;
+
+    #[derive(Default)]
+    struct FailpointsHook {
+        pre_append_action: Arc<RwLock<HashMap<u64, Box<Action>>>>,
+        pre_append_timer: AtomicU64,
+    }
+
+    impl FailpointsHook {
+        fn new() -> Self {
+            Self {
+                pre_append_action: Arc::new(RwLock::new(HashMap::default())),
+                pre_append_timer: AtomicU64::new(0),
+            }
+        }
+
+        fn register_pre_append_action(
+            &mut self,
+            index: u64,
+            action: impl FnOnce() + Send + Sync + 'static,
+        ) {
+            self.pre_append_action
+                .write()
+                .insert(index, Box::new(action));
+        }
+    }
+
+    impl EventListener for FailpointsHook {
+        fn pre_append(&self) {
+            let idx = self
+                .pre_append_timer
+                .fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+            if let Some(action) = self.pre_append_action.write().remove(&idx) {
+                action();
+            }
+        }
+    }
+
+    #[test]
+    fn test_concurrent_write_error() {
+        // truncate and sync when write error
+        let dir = tempfile::Builder::new()
+            .prefix("handle_io_error")
+            .tempdir()
+            .unwrap();
+        let cfg = Config {
+            dir: dir.path().to_str().unwrap().to_owned(),
+            bytes_per_sync: ReadableSize::kb(1024),
+            target_file_size: ReadableSize::kb(1024),
+            batch_compression_threshold: ReadableSize(0),
+            ..Default::default()
+        };
+
+        // b0 (ctx); b1 success; b2 fail, truncate; b3 success
+        let mut hook = FailpointsHook::new();
+        hook.register_pre_append_action(2, || {
```

Comment:
I don't particularly like this approach. Could you try using `cfg_callback`? might be easier.